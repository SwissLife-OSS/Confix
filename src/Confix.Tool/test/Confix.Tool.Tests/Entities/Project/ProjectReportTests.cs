using Confix.Inputs;
using Confix.Tool.Commands.Temp;
using Confix.Tool.Schema;
using Confix.Utilities;
using Moq;

namespace Confix.Entities.Component.Configuration.Middlewares;

public class ProjectReportTests : IAsyncLifetime
{
    private readonly Mock<IGitService> _git;
    private readonly TestConfixCommandline _cli;

    public ProjectReportTests()
    {
        _git = new Mock<IGitService>();
        _cli = new TestConfixCommandline(x => x.AddTestService<IGitService>(_ => _git.Object));
    }

    [Fact]
    public async Task Should_OutputReportToCli_When_NoFileSpecified()
    {
        // arrange
        _cli.Directories.Home.CreateConfixRc(_confixRc);
        _cli.Directories.Content
            .CreateConfixProject(path: $"src/project/{FileNames.ConfixProject}");
        _cli.Directories.Content.CreateConfixSolution();
        _cli.ExecutionContext = _cli.ExecutionContext with
        {
            CurrentDirectory = _cli.Directories.Content
                .Append("src")
                .Append("project")
        };

        _git.SetupGitRoot(_cli.Directories.Content);
        _git.SetupGitBranch();
        _git.SetupGitTags();
        _git.SetupGitRepoInfo();

        // act
        await _cli.RunAsync("project report");

        // assert
        SnapshotBuilder.New().AddOutput(_cli).MatchSnapshot();
    }

    /// <inheritdoc />
    public Task InitializeAsync()
    {
        return Task.CompletedTask;
    }

    /// <inheritdoc />
    public Task DisposeAsync()
    {
        _cli.Dispose();
        return Task.CompletedTask;
    }

    private const string _confixRc = """
            {
                "isRoot": true,
                "component":
                {
                     "inputs": [
                          {
                            "type": "graphql"
                          },
                          {
                            "type": "dotnet"
                          }
                        ]
                },
                "project": {
                     "componentProviders": [
                          {
                            "name": "dotnet-package",
                            "type": "dotnet-package"
                          }
                        ]
                }
            }
        """;
}

file static class Extensions
{
    public static void SetupGitRoot(this Mock<IGitService> service, FileSystemInfo root)
    {
        service.Setup(
                x => x.GetRootAsync(It.IsAny<GitGetRootConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => root.FullName);
    }

    public static void SetupGitBranch(
        this Mock<IGitService> service,
        string branch = "master")
    {
        service.Setup(
                x => x.GetBranchAsync(It.IsAny<GitGetBranchConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => branch);
    }

    public static void SetupGitTags(
        this Mock<IGitService> service,
        string[]? tags = default!)
    {
        tags ??= new[]
        {
            "1.0.0",
            "1.1.0"
        };
        service.Setup(
                x => x.GetTagsAsync(It.IsAny<GitGetTagsConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => tags);
    }

    public static void SetupGitRepoInfo(
        this Mock<IGitService> service,
        string hash = "1234567890",
        string message = "Initial commit",
        string author = "John Doe",
        string email = "jd@gmail.com")
    {
        service.Setup(
                x => x.GetRepoInfoAsync(It.IsAny<GitGetRepoInfoConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => new GitGetRepoInfoResult(hash, message, author, email));
    }
}
