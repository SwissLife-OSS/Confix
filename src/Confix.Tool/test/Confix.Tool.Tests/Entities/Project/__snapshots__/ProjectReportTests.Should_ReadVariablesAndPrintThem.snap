--------------------------------------------------
### CLI Output
--------------------------------------------------
  Configuration files of type .confixrc located at <</tmp>>/home/.confixrc
  Configuration files of type .confix.solution located at <</tmp>>/content/.confix.solution
  Configuration files of type .confix.solution located at <</tmp>>/content/src/project/.confix.project
  Replaced magic path $project:/test.json with <</tmp>>/content/src/project/test.json.
  Replaced magic path $project:/vars.json with <</tmp>>/content/src/project/vars.json.
  Merged .confix.solution from .confixrc
  Merged .confix.project from .confix.solution
  Merged .confix.component from .confix.solution
✓ Running in scope Project
✓ Configuration loaded
  Found a inline configuration file '<</tmp>>/content/src/project/test.json'
✓ Active Environment is prod
✓ Component inputs loaded
✓ Dependency providers loaded
  Commit: 1234567890 by John Doe on master Initial commit
  Repository: <</tmp>> <</tmp>>/content
  Solution: content .
  Project: src.project src/project
  Start loading components from project 'src.project'
✓ Loaded 0 components
  Detected 3 variables
  Resolving 3 variables from foo
  Report: <</tmp>>/content/src/project/test.json prod
[
   {
      "configurationPath": "<</tmp>>/content/src/project/test.json",
      "environment": "prod",
      "timestamp": <<date>>,
      "project": {
         "name": "src.project",
         "path": "src/project"
      },
      "solution": {
         "name": "content",
         "path": "."
      },
      "repository": {
         "name": "<</tmp>>",
         "originUrl": "https://foobar.com"
      },
      "commit": {
         "hash": "1234567890",
         "message": "Initial commit",
         "author": "John Doe",
         "email": "jd@gmail.com",
         "branch": "master",
         "tags": [
            "1.0.0",
            "1.1.0"
         ]
      },
      "variables": [
         {
            "providerName": "foo",
            "providerType": "local",
            "name": "test.bool",
            "hash": "B5BEA41B6C623F7C09F1BF24DCAE58EBAB3C0CDD90AD966BC43A45B44867E12B",
            "path": "/root/nested/bool"
         },
         {
            "providerName": "foo",
            "providerType": "local",
            "name": "test.string",
            "hash": "9F86D081884C7D659A2FEAA0C55AD015A3BF4F1B2B0B822CD15D6C15B0F00A08",
            "path": "/root/nested/string"
         },
         {
            "providerName": "foo",
            "providerType": "local",
            "name": "test.object",
            "hash": "7A38BF81F383F69433AD6E900D35B3E2385593F76A7B7AB5D4355B8BA41EE24B",
            "path": "/root/nested/object"
         }
      ],
      "components": [
      ],
      "dependencies": [
      ]
   }
]
